ambry-frontend/src/main/java/com/github/ambry/frontend/s3/S3BatchDeleteHandler.java

```diff
@@ -0,0 +1,174 @@
+   */
+  private class BatchDeleteCallback implements Callback<Long> {
+    private final RetainingAsyncWritableChannel channel;
+    private final RestRequest restRequest;
+    private final DeleteBlobHandler deleteBlobHandler;
+
+    public BatchDeleteCallback(RetainingAsyncWritableChannel channel, RestRequest restRequest, DeleteBlobHandler deleteBlobHandler) {
+      this.channel = channel;
+      this.restRequest = restRequest;
+      this.deleteBlobHandler = deleteBlobHandler;
+    }
+
+    public void onDeleteCompletion(boolean success, String key) {
+      if (success) {
+        deleted.add(key);
+      }
+      else {
+        errors.add(key);
+      }
+    }
+
+    @Override
+    public void onCompletion(Long respLong, Exception exception) {
+      if (exception == null) {
+        // Data read successfully
+        try {
+          // Get the retained content from the channel
+          ByteBuf byteBuffer = channel.consumeContentAsByteBuf();
+          // TODO byte array?
+          // Convert ByteBuf to byte array
+          byte[] byteArray = new byte[byteBuffer.readableBytes()];
+          byteBuffer.readBytes(byteArray);
+
+          // Deserialize into S3BatchDeleteObjects
+          XmlMapper xmlMapper = new XmlMapper();
+          S3MessagePayload.S3BatchDeleteObjects deleteRequest = xmlMapper.readValue(byteArray, S3MessagePayload.S3BatchDeleteObjects.class);
+
+          // Process each delete key in the batch
+          for (S3MessagePayload.S3BatchDeleteKeys object : deleteRequest.getObjects()) {
+            RequestPath requestPath = (RequestPath) restRequest.getArgs().get(RestUtils.InternalKeys.REQUEST_PATH);
+
+            // Construct the delete path
+            // TODO: confirm that getPathAfterPrefixes is indeed "/named/application/container"
+            String singleDeletePath = requestPath.getPathAfterPrefixes() + "/" + object.getKey();
+
+            // Create a new RequestPath for the delete operation
+            List<String> emptyList = new ArrayList<>();
+            RequestPath newRequestPath = RequestPath.parse(singleDeletePath, restRequest.getArgs(), emptyList, requestPath.getClusterName());
+            WrappedRestRequest singleDeleteRequest = new WrappedRestRequest(restRequest);
+            singleDeleteRequest.setArg(RestUtils.InternalKeys.REQUEST_PATH, newRequestPath);
+
+            CountDownLatch latch = new CountDownLatch(1);
+            NoOpResponseChannel noOpResponseChannel = new NoOpResponseChannel();
+
+            deleteBlobHandler.handle(singleDeleteRequest, noOpResponseChannel, new Callback<Void>() {
+              @Override
+              public void onCompletion(Void result, Exception exception) {
+                  // Call our custom onDeleteCompletion to track success/failure
+                  boolean success = exception == null;
+                  onDeleteCompletion(success, object.getKey());
+                  latch.countDown();
+                }
```

Comment:
We also want to differentiate the error cases. Some keys failed with NotFound, some keys failed with InternalServerError etc. Let's keep a map from key to its delete status?